### DISCUSSION TITLE
From flask to emmett

### DISCUSSION THREAD

**Discussion by josejachuf** [emmett-framework/emmett#373] [Q&A] [answered]
Hello, 
I am seeing an example in Flask and uses **make_response**. I do not really know what he does. 

in Flask:
```
response = make_response(X)
response.headers['Content-Type'] = "application/octet-stream"
return response
```

In Emmett I made this:
```
from emmett import response
...
response.headers['Content-Type'] = "application/octet-stream"
return X
```

It's right? 

Best Regards
Jose

**Comment by gi0baro** [comment]
@josejachuf it depends. Are you trying to serve a file? a stream of binary data?

**Reply by josejachuf** [reply]
I think both. Specifically it is a vector tile (*.pbf). These are created and cached as files. If the file exists, that file is served, but it is found to the database and cache for later use.

**Comment by josejachuf** [comment, marked as answer]
**tile** is bytes type (**<class 'bytes'>**) In Flask this works fine
    
    tile = get_tile(z, x, y)
    response = make_response(tile)
    response.headers['Content-Type'] = "application/octet-stream"
    return response


in emmett I made this:

    tile = get_tile(z, x, y)
    response.headers['Content-Type'] = "application/octet-stream" 
    return tile


This returns a bytes too, but the maps viewer shows an error and does not render the map

**Reply by josejachuf** [reply]
It was missing to define the type of output:

> output='bytes'

It was resolved 

Best regards
Jose

**Reply by gi0baro** [reply]
@josejachuf sorry for the delay. Defining the `output` shouldn't be necessary, in fact its design was for performance reasons only. Then probably I should enhance the automatic output recognition.

Anyway, that works, but probably is not what you really want. In case you have an accessible file, then `stream_file` from `emmett.helpers` might help you (avoid to load the entire file contents in RAM). You can even use lower APIs, like raising an `emmett.http.HTTPIO` exception from your route, but without further information about what actually `X` is in your code, this is the best advice I can give you.

**Reply by josejachuf** [reply]
Hi @gi0baro, thanks for your answare.
X is a Vector Tile [1], [2]. They are very small files. I am going to serve you from the cache directly with Nginx, without going through Emmett. When they do not exist, Emmett will be created. 


[1] https://en.wikipedia.org/wiki/Vector_tiles
[2] https://postgis.net/docs/ST_AsMVT.html
